src/main/java/org/openstreetmap/atlas/checks/validation/intersections/EdgeCrossingEdgeCheck.java

```diff
@@ -191,4 +119,87 @@ public boolean validCheckForObject(final AtlasObject object)
+                        && !HighwayTag.CROSSING.equals(highwayTag)
+                        && highwayTag.isMoreImportantThanOrEqualTo(this.minimumHighwayType);
+            }
+        }
+        return false;
+    }
+
+    /**
+     * A {@link Function} for an {@link EdgeWalker} that collects all connected invalid crossings.
+     *
+     * @return a {@link Set} of invalid crossing {@link Edge}s
+     */
+    private Function<Edge, Stream<Edge>> getInvalidCrossingEdges()
+    {
+        return edge ->
+        {
+            // Prepare the edge being tested for checks
+            final PolyLine edgeAsPolyLine = edge.asPolyLine();
+            final Rectangle edgeBounds = edge.bounds();
+            // If layer tag is present use its value, else use the OSM default
+            final Optional<Long> edgeLayer = Validators.hasValuesFor(edge, LayerTag.class)
+                    ? LayerTag.getTaggedValue(edge) : Optional.of(OSM_LAYER_DEFAULT);
+
+            // Retrieve crossing edges
+            final Atlas atlas = edge.getAtlas();
+            return Iterables
+                    .asList(atlas.edgesIntersecting(edgeBounds,
+                            // filter out the same edge and non-valid crossing edges
+                            crossingEdge -> edge.getIdentifier() != crossingEdge.getIdentifier()
+                                    && this.isValidCrossingEdge(crossingEdge)))
+                    .stream()
+                    // Go through crossing items and collect invalid crossings
+                    // NOTE: Due to way sectioning same OSM way could be marked multiple times here.
+                    // However,
+                    // MapRoulette will display way-sectioned edges in case there is an invalid
+                    // crossing.
+                    // Therefore, if an OSM way crosses another OSM way multiple times in separate
+                    // edges,
+                    // then each edge will be marked explicitly.
+                    .filter(crossingEdge ->
+                    {
+                        final PolyLine crossingEdgeAsPolyLine = crossingEdge.asPolyLine();
+                        final Optional<Long> crossingEdgeLayer = Validators
+                                .hasValuesFor(crossingEdge, LayerTag.class)
+                                        ? LayerTag.getTaggedValue(crossingEdge)
+                                        : Optional.of(OSM_LAYER_DEFAULT);
+                        return edgeAsPolyLine.intersections(crossingEdgeAsPolyLine).stream()
+                                .anyMatch(intersection -> !canCross(edgeAsPolyLine, edgeLayer,
+                                        crossingEdgeAsPolyLine, crossingEdgeLayer, intersection));
+                    });
+        };
+    }
+
+    /**
+     * A simple {@link EdgeWalker} that filters using nextCandidates.
+     */
+    private class EdgeCrossingEdgeWalker extends EdgeWalker
```

Comment:
The `EdgeWalker` I am using is really basic, does it make sense to add something like this as a class in atlas? It seems like a lot of work each time to make a new class when you just need to pass a `nextCandidates` function and a maybe a `candidateFilter`.